src/Xamarin.Android.Build.Tasks/Tasks/GenerateCompressedAssembliesNativeSourceFiles.cs

```diff
@@ -0,0 +1,86 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.Build.Framework;
+
+namespace Xamarin.Android.Tasks
+{
+	public class GenerateCompressedAssembliesNativeSourceFiles : AndroidTask
+	{
+		public override string TaskPrefix => "GCANSF";
+
+		[Required]
+		public ITaskItem[] ResolvedAssemblies { get; set; }
+
+		[Required]
+		public string [] SupportedAbis { get; set; }
+
+		[Required]
+		public string EnvironmentOutputDirectory { get; set; }
+
+		[Required]
+		public bool Debug { get; set; }
+
+		[Required]
+		public bool EnableCompression { get; set; }
+
+		public override bool RunTask ()
+		{
+			GenerateCompressedAssemblySources ();
+			return !Log.HasLoggedErrors;
+		}
+
+		void GenerateCompressedAssemblySources ()
+		{
+			if (Debug || !EnableCompression) {
+				Generate (null);
+				return;
+			}
+
+			var assemblies = new SortedDictionary<string, CompressedAssemblyInfo> (StringComparer.Ordinal);
+			foreach (ITaskItem assembly in ResolvedAssemblies) {
+				if (bool.TryParse (assembly.GetMetadata ("AndroidSkipAddToPackage"), out bool value) && value) {
+					continue;
+				}
+
+				if (assemblies.ContainsKey (assembly.ItemSpec)) {
+					continue;
+				}
+
+				var fi = new FileInfo (assembly.ItemSpec);
+				if (!fi.Exists) {
+					Log.LogError ($"Assembly {assembly.ItemSpec} does not exist");
+					continue;
+				}
+
+				assemblies.Add (Path.GetFileName (assembly.ItemSpec), new CompressedAssemblyInfo ((uint)fi.Length));
```

Comment:
What should happen when we get an assembly > 4GB in size? Should this be `checked((uint) fi.Length)`?

Reply:
In theory yes, we should check it. In practice, we won't be able to put such a file in the APK - even compressed. This is because ZIP file size fields for both uncompressed and compressed files are 32-bit, unless we use ZIP64 extensions, which APKs don't. So if we want to enforce this limit somewhere, I think it would be good to add a general warning about files too big to be placed in the APK somewhere else in the build tasks.

Reply:
Part of the concern here is that while, *yes*, this isn't possible, *how* does it fail?  What is the error message produced when someone tries to add a 5GB assembly to the `.apk`?  Is that error message *useful* and actionable?

I have no idea what we *currently* do.  Perhaps what we already do is sufficient.

Reply:
If we do anything at all, I think it should be done elsewhere. This code runs very late in the process, when building the APK - we should warn about it earlier than that.